hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGeneratorBase.java

```diff
@@ -0,0 +1,147 @@
+
+import static org.apache.hudi.common.config.LockConfiguration.DEFAULT_LOCK_ACQUIRE_CLIENT_RETRY_WAIT_TIME_IN_MILLIS;
+import static org.apache.hudi.common.config.LockConfiguration.DEFAULT_LOCK_ACQUIRE_NUM_RETRIES;
+import static org.apache.hudi.common.config.LockConfiguration.DEFAULT_LOCK_ACQUIRE_WAIT_TIMEOUT_MS;
+import static org.apache.hudi.common.config.LockConfiguration.LOCK_ACQUIRE_CLIENT_NUM_RETRIES_PROP_KEY;
+import static org.apache.hudi.common.config.LockConfiguration.LOCK_ACQUIRE_CLIENT_RETRY_WAIT_TIME_IN_MILLIS_PROP_KEY;
+import static org.apache.hudi.common.config.LockConfiguration.LOCK_ACQUIRE_WAIT_TIMEOUT_MS_PROP_KEY;
+
+/**
+ * Base time generator facility that maintains lock-related utilities.
+ */
+public abstract class TimeGeneratorBase implements TimeGenerator, Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TimeGeneratorBase.class);
+
+  /**
+   * The lock provider.
+   */
+  private volatile LockProvider lockProvider;
+  /**
+   * The maximum times to retry in case there are failures.
+   */
+  private final int maxRetries;
+  /**
+   * The maximum time to wait for each time generation to resolve the clock skew issue on distributed hosts.
+   */
+  private final long maxWaitTimeInMs;
+  /**
+   * The maximum time to block for acquiring a lock.
+   */
+  private final int lockAcquireWaitTimeInMs;
+
+  protected final HoodieTimeGeneratorConfig config;
+  private final LockConfiguration lockConfiguration;
+
+  /**
+   * The hadoop configuration.
+   */
+  private final SerializableConfiguration hadoopConf;
+
+  public TimeGeneratorBase(HoodieTimeGeneratorConfig config, SerializableConfiguration hadoopConf) {
+    this.config = config;
+    this.lockConfiguration = config.getLockConfiguration();
+    this.hadoopConf = hadoopConf;
+
+    maxRetries = lockConfiguration.getConfig().getInteger(LOCK_ACQUIRE_CLIENT_NUM_RETRIES_PROP_KEY,
+        Integer.parseInt(DEFAULT_LOCK_ACQUIRE_NUM_RETRIES));
+    lockAcquireWaitTimeInMs = lockConfiguration.getConfig().getInteger(LOCK_ACQUIRE_WAIT_TIMEOUT_MS_PROP_KEY,
+        Integer.parseInt(DEFAULT_LOCK_ACQUIRE_WAIT_TIMEOUT_MS));
+    maxWaitTimeInMs = lockConfiguration.getConfig().getLong(LOCK_ACQUIRE_CLIENT_RETRY_WAIT_TIME_IN_MILLIS_PROP_KEY,
+        Long.parseLong(DEFAULT_LOCK_ACQUIRE_CLIENT_RETRY_WAIT_TIME_IN_MILLIS));
+  }
+
+  protected LockProvider getLockProvider() {
+    // Perform lazy initialization of lock provider only if needed
+    if (lockProvider == null) {
+      synchronized (this) {
+        if (lockProvider == null) {
+          String lockProviderClass = lockConfiguration.getConfig().getString("hoodie.write.lock.provider");
+          LOG.info("LockProvider for TimeGenerator: " + lockProviderClass);
+          lockProvider = (LockProvider) ReflectionUtils.loadClass(lockProviderClass,
+              lockConfiguration, hadoopConf.get());
+        }
+      }
+    }
+    return lockProvider;
+  }
+
+  public void lock() {
```

Comment:
is there a common place to move this logic to? I guess this would be common to all locking code, right>?

Reply:
Yeah, I think we can have a refactoring to the failure retries code snippet, maybe we can abstrac it out into a utilitiy method.